_posts/2020-05-18-microservice-architectures.md

```diff
@@ -0,0 +1,80 @@
+
+Microservices have a [long history](https://www.researchgate.net/profile/Mohammad_Hadi_Valipour/publication/224588022_A_Brief_Survey_of_Software_Architecture_Concepts_and_Service_Oriented_Architecture/links/0a85e5346344791494000000/A-Brief-Survey-of-Software-Architecture-Concepts-and-Service-Oriented-Architecture.pdf), with roots in SOA, and [distributed computing](https://en.wikipedia.org/wiki/Distributed_computing#History).  Microservices is an emotive subject, with arguments both [for](https://martinfowler.com/articles/dont-start-monolith.html) and [against](https://martinfowler.com/bliki/MonolithFirst.html) building microservices from the start.  
+
+But how do we characterise a microservice?  We do not characterise based on the number of lines of code, but rather on [bounded context](https://martinfowler.com/bliki/BoundedContext.html).  
+
+For me, microservices:
+* Are event driven
+* Are domain oriented
+* Decentralise data management
+
+## Event Driven
+
+Serverless computing is event driven at it's core.  Whether invoking [functions](https://docs.aws.amazon.com/lambda/latest/dg/lambda-invocation.html), [consuming services](https://docs.aws.amazon.com/rekognition/latest/dg/stored-video-lambda.html) or [building integrations](https://aws.amazon.com/eventbridge/).  
+
+Events comprise of two core elements:
+* Domain event - the important detail of the event and includes important attributes such as timestamp, and eventType.  We use the [cloud event specification](https://cloudevents.io/) for defining domain events.
+* Envelope - wraps the domain event and captures details of the event transport, such as [SNS](https://docs.aws.amazon.com/sns/latest/dg/sns-message-and-json-formats.html) or [EventBridge](https://docs.aws.amazon.com/eventbridge/latest/userguide/aws-events.html) events.
+
+![clean architecture diagram showing onion of layers including handler, integration with ports and adapters, service and domain layers ](/img/microservice-architectures/clean-architecture.png)
+
+By following [clean architecture](https://blog.cleancoder.com/uncle-bob/2012/08/13/the-clean-architecture.html) principles, separating envelope unmarshalling from acting on domain events, provides additional flexibility on your deployment architecture.  For example, you can use the same domain processing logic to handle synchronous requests sent via API gateway or asynchronously via SNS.
```

Comment:
_Following clean architecture principles (such as separating envelope unmarshalling from acting on domain events) provides additional flexibility in your deployment architecture._

??